agent/consul/gateways/controller_gateways.go

```diff
@@ -2,27 +2,166 @@ package gateways
 
 import (
 	"context"
-
-	"github.com/hashicorp/go-hclog"
-
+	"errors"
+	"fmt"
 	"github.com/hashicorp/consul/agent/consul/controller"
 	"github.com/hashicorp/consul/agent/consul/fsm"
+	"github.com/hashicorp/consul/agent/consul/gateways/datastore"
 	"github.com/hashicorp/consul/agent/consul/state"
 	"github.com/hashicorp/consul/agent/consul/stream"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/go-hclog"
 )
 
 type apiGatewayReconciler struct {
-	fsm    *fsm.FSM
 	logger hclog.Logger
+	store  datastore.DataStore
+}
+
+func (r apiGatewayReconciler) retrieveAllRoutesFromStore() ([]structs.BoundRoute, error) {
+	tcpRoutes, err := r.store.GetConfigEntriesByKind(structs.TCPRoute)
+	if err != nil {
+		return nil, err
+	}
+
+	//TODO not implemented
+	//httpRoutes, err := r.store.GetConfigEntriesByKind(structs.HTTPRoute)
+	//if err != nil {
+	//	return nil, err
+	//}
+
+	routes := []structs.BoundRoute{}
+	for _, r := range tcpRoutes {
+		if r == nil {
+			continue
+		}
+		routes = append(routes, r.(*structs.TCPRouteConfigEntry))
+	}
+	//TODO not implemented
+	//for _, r := range httpRoutes {
+	//	routes = append(routes, r.(*structs.HTTPRouteConfigEntry))
+	//}
+	fmt.Println(routes)
+	return routes, nil
 }
 
 func (r apiGatewayReconciler) Reconcile(ctx context.Context, req controller.Request) error {
+	//TODO thomases wrapper function
+	entry, err := r.store.GetConfigEntry(req.Kind, req.Name, req.Meta)
+	if err != nil {
+		return err
+	}
+
+	if entry == nil {
+		r.logger.Error("cleaning up deleted gateway object", "request", req)
+		if err := r.store.Delete(&structs.BoundAPIGatewayConfigEntry{
+			Kind:           structs.BoundAPIGateway,
+			Name:           req.Name,
+			EnterpriseMeta: *req.Meta,
+		}); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	gatewayEntry := entry.(*structs.APIGatewayConfigEntry)
+	//TODO is this what needs to happen for the validation step
+	err = gatewayEntry.Validate()
+	if err != nil {
+		r.logger.Debug("persisting gateway status", "gateway", gatewayEntry)
+		if err := r.store.UpdateStatus(gatewayEntry); err != nil {
+			return err
+		}
+		return err
```

Comment:
If the gateway entry fails to validate and returns an error, then the status update fails too, you will lose the context of the validation error when you return.

I might suggest doing this instead:

```go
if updateErr := r.store.UpdateStatus(gatewayEntry); updateErr != nil {
    return fmt.Errorf("%v: %v", err, updateErr)
}
return err
```